### DISCUSSION TITLE
Setup mode

### DISCUSSION THREAD

**Discussion by n8mgr** [SiaFoundation/hostd#825] [Feature Requests]
If the user has not configured a seed or API password, `hostd` should enter a setup mode with a single endpoint that allows the user to configure the yaml file via the web UI instead of failing to start. After the user has completed the initial setup, hostd should start normally without needing to be restarted.

**Comment by n8mgr** [comment]
@ChrisSchinnerl @peterjan @alexfreska this is a feature I would like to discuss. We have this for the desktop apps, but configuring the apps on headless devices is still a friction point.

**Comment by ChrisSchinnerl** [comment]
That's interesting. Do you want to spin up the regular UI or a special page only for settings and submitting the config after which it gets written to disk and the actual UI and hostd get spin up?

**Comment by alexfreska** [comment]
The way I understand:

Pre setup:
- The daemon starts with 1 /api/setup endpoint.
- Also serves the regular UI as normal.
- The daemon can open the UI to a special route localhost:9980/setup.
  - This page is part of the regular UI but only interacts only with the /api/setup endpoint.

Post setup:
- On successful submission to /api/setup:
- The setup UI page closes itself.
- The daemon spins up the full API, without /api/setup.
- The daemon opens the normal UI root page localhost:9980/setup.
- Manually visiting the UI setup page would just 404/redirect back to root.


Is this in line with your thinking? Do you both envision this is as only for first time config? Or could the user visit this page post setup and modify the config/trigger restart?

**Comment by n8mgr** [comment]
That’s pretty much exactly the workflow I envisioned. A one time API and setup flow that disappears after it’s complete and writes a yaml config to disk. The regular UI will be redirected to a setup page since the user can’t login.

We could add a config endpoint to the regular API to update the YAML, but it should not expose the wallet seed.

**Comment by ChrisSchinnerl** [comment]
Since serving the regular UI doesn't make sense during the setup stage, should we just serve the setup page on `localhost:9980`? Then users can just visit the same URL they usually would without having to add the extra `/setup`. I imagine that is a bit nicer for when the daemon can't auto-open a browser.

**Comment by n8mgr** [comment]
@alexfreska would creating a distinct UI overcomplicate the UI packaging and build process? Having one external UI package is already a bit of a pain.

**Comment by alexfreska** [comment]
We can make the setup load on `localhost:9980` (or at least instantly redirect etc). Yeah I would really want to make it part of the current UI so that we don't add more complexity to matching up all the versions and dependencies - this shouldn't prevent us from doing anything we need, including making it feel separate or loading on the same route etc.

**Comment by pcfreak30** [comment]
IMO if you do this on `hostd`, should probably apply it to `renterd` as well, 🤷‍♂ .
